File: cpp/714.best-time-to-buy-and-sell-stock-with-transaction-fee.cpp
/*
 * @lc app=leetcode id=714 lang=cpp
 *
 * [714] Best Time to Buy and Sell Stock with Transaction Fee
 */
#include <vector>
#include <climits>
using namespace std;
// @lc code=start
class Solution
{
public:
    int maxProfit(vector<int> &prices, int fee)
    {
        int profit = 0;
        int cost = INT_MAX;
        for (auto p : prices)
        {
            if (p > cost)
            {
                profit += p - cost;
                cost = p;
            }
            else if (p + fee < cost)
            {
                cost = p + fee;
            }
        }
        return profit;
    }
};
// @lc code=end

File: py/714.best-time-to-buy-and-sell-stock-with-transaction-fee.py
#
# @lc app=leetcode id=714 lang=python3
#
# [714] Best Time to Buy and Sell Stock with Transaction Fee
#

# @lc code=start
class Solution:
    def maxProfit(self, prices: List[int], fee: int) -> int:
        profit = 0
        cost = 1000000
        for p in prices:
            if p > cost:
                profit += p-cost
                cost = p
            elif p+fee < cost:
                cost = p+fee
        return profit

# @lc code=end
